core-rust/state-manager/src/store/parent_map.rs

```diff
@@ -0,0 +1,315 @@
+    /// trees already-existing in the given [`SubstateNodeMetadataStore`].
+    /// API note: the results are grouped by [`SubstateNodeMetadata`] (since it may be shared by
+    /// many [`NodeId`]s).
+    pub fn batch_resolve<S: SubstateNodeMetadataStore>(
+        metadata_store: &S,
+        substate_changes: &[SubstateChange],
+    ) -> impl Iterator<Item = (Vec<NodeId>, SubstateNodeMetadata)> {
+        // Index the new Nodes by their parent Substates (using `IndexedScryptoValue`):
+        let changed_nodes_by_parent = substate_changes
+            .iter()
+            .filter_map(|substate_change| match &substate_change.action {
+                ChangeAction::Create(bytes) | ChangeAction::Update(bytes) => {
+                    let owned_node_ids = IndexedScryptoValue::from_slice(bytes).unwrap().unpack().1;
+                    if owned_node_ids.is_empty() {
+                        return None;
+                    }
+                    let parent_reference = SubstateReference(
+                        substate_change.node_id,
+                        substate_change.partition_number,
+                        substate_change.substate_key.clone(),
+                    );
+                    Some((parent_reference, owned_node_ids))
+                }
+                ChangeAction::Delete => None,
+            })
+            .collect::<IndexMap<_, _>>();
+
+        // Invert the multimap (to obtain a Node -> parent Substate map):
+        let changed_parent_by_node = changed_nodes_by_parent
+            .iter()
+            .flat_map(|(parent, children)| children.iter().map(|child| (*child, parent.clone())))
+            .collect::<NonIterMap<_, _>>();
+
+        // Find the parents which do _not_ know their parent within the `SubstateChange`s:
+        let top_level_parent_node_ids = changed_nodes_by_parent
+            .keys()
+            .filter(|parent| !changed_parent_by_node.contains_key(&parent.0))
+            .map(|parent| parent.0)
+            .collect::<Vec<_>>();
+
+        // Load the missing metadata of these parents from the `SubstateNodeMetadataStore`:
+        let existing_metadatas = metadata_store.batch_get_metadata(&top_level_parent_node_ids);
+
+        // Construct a Node -> root map needed for the `NodeMetadataResolver` helper structure:
+        let existing_root_by_node = top_level_parent_node_ids
+            .into_iter()
+            .zip(existing_metadatas)
+            .filter_map(|(parent, existing_metadata)| {
+                existing_metadata.map(|metadata| (parent, metadata.root))
+            })
+            .collect::<NonIterMap<_, _>>();
+
+        let resolver = Self {
+            changed_parent_by_node,
+            existing_root_by_node,
+        };
+
+        // Simply query the fully-preloaded `NodeMetadataResolver` for each Node:
+        changed_nodes_by_parent
+            .into_iter()
+            .map(move |(parent, child_node_ids)| {
```

Comment:
This is O(N^2) in the worst case without a cache, or more technically O(N * D) for D the max depth, which I think can be D ~ N for nested key value stores.

Reply:
I don't know what your thoughts are / whether you think it's worth adding a cache to reduce it to O(N)? Might not be worthwhile in practice, happy for you to make that call

Reply:
Cool idea, I think I'll be able to neatly integrate the cache 👍

Reply:
As discussed on our call, I went for a more proper "propagate the topmost parents down" approach (which is definitely O(N)).

Achieving it without recursion (and without rewriting `Vec<>`s until it is not O(N) anymore) was a challenge, and resulted in a less-then-perfectly-readable code, so I added heavy comments and extra unit tests. PTAL.

All the other rename requests are already applied.
The other optimization (https://github.com/radixdlt/babylon-node/pull/511#discussion_r1235023152) happens in next PR.